File: code401challenges/src/main/java/code401challenges/hastable/Hashtable.java
package code401challenges.hastable;

public class Hashtable {
    //instance variables
    int size;
    HashEntry[] map;

    //constructor function
    public Hashtable (int size) {
        this.size = size;
        this.map = new HashEntry[size];
    }


    //instance methods
    public void add (String key, String value) {
        int idx = hash(key);
        HashEntry hashEntry = new HashEntry(key, value);

        if (map[idx] == null){
            map[idx] = hashEntry;
        } else {
            HashEntry current = map[idx];
            while (current.next != null) {
                current = current.next;
            }
            current.next = hashEntry;
        }
    }

    public String get (String key) {

        return "";
    }

    public Boolean contains (String key) {
        int idx = hash(key);

        return false;
    }

    public int hash (String key) {
        return key.hashCode() % this.size;
    }


}

File: js401challenges/HashTable/HashTable.js
'use strict';

let LinkedList = require('../LinkedList/LinkedList.js');

class HashTable {

  constructor (size) {
    this.size = size;
    this.map = new Array(size).fill();
  }

  hash(key) {
    let chars = key.split('');
    return chars.reduce((p, character) => {
      return p + character.charCodeAt(0);
    }, 0) * 599 % this.size;
  }

  add (key, value) {
    let hashCode = this.hash(key);

    if (!this.map[hashCode]) {
      this.map[hashCode] = new LinkedList();
    }

    let entry = { [key]: value };
    this.map[hashCode].add(entry);
  }

  get (key) {
    let hashCode = this.hash(key);
    if (!this.map[hashCode]) { return null; }

    let current = this.map[hashCode].head;
    while (current) {
      let data = current.value;
      if (data[key]) { return data[key]; }
      current = current.next;
    }

    return null;
  }

  contains (key) {
    let hashCode = this.hash(key);
    if (!this.map[hashCode]) { return false; }

    let current = this.map[hashCode].head;
    while (current) {
      let data = current.value;
      if (data[key]) { return true; }
      current = current.next;
    }

    return false;
  }

}

module.exports = HashTable;
